pkg/controller/common/watches/state.go

```diff
@@ -7,25 +7,26 @@ package watches
 import (
 	corev1 "k8s.io/api/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 // NewDynamicWatches creates an initialized DynamicWatches container.
 func NewDynamicWatches() DynamicWatches {
 	return DynamicWatches{
-		ConfigMaps:          NewDynamicEnqueueRequest[*corev1.ConfigMap](),
-		Secrets:             NewDynamicEnqueueRequest[*corev1.Secret](),
-		Services:            NewDynamicEnqueueRequest[*corev1.Service](),
-		Pods:                NewDynamicEnqueueRequest[*corev1.Pod](),
-		ReferencedResources: NewDynamicEnqueueRequest[client.Object](),
+		ConfigMaps:          NewDynamicEnqueueRequest[*corev1.ConfigMap, reconcile.Request](),
+		Secrets:             NewDynamicEnqueueRequest[*corev1.Secret, reconcile.Request](),
+		Services:            NewDynamicEnqueueRequest[*corev1.Service, reconcile.Request](),
+		Pods:                NewDynamicEnqueueRequest[*corev1.Pod, reconcile.Request](),
+		ReferencedResources: NewDynamicEnqueueRequest[client.Object, reconcile.Request](),
```

Comment:
You don't need the extra type parameter if it is always going to be a `reconcile.Request`
```diff
diff --git a/pkg/controller/association/dynamic_watches.go b/pkg/controller/association/dynamic_watches.go
index 66a93cd48..bc0927291 100644
--- a/pkg/controller/association/dynamic_watches.go
+++ b/pkg/controller/association/dynamic_watches.go
@@ -132,7 +132,7 @@ func (r *Reconciler) reconcileWatches(ctx context.Context, associated types.Name
 func reconcileGenericWatch[T client.Object, R reconcile.Request](
 	associated types.NamespacedName,
 	associations []commonv1.Association,
-	dynamicRequest *watches.DynamicEnqueueRequest[T, reconcile.Request],
+	dynamicRequest *watches.DynamicEnqueueRequest[T],
 	watchName string,
 	watchedFunc func() ([]types.NamespacedName, error),
 ) error {
@@ -158,7 +158,7 @@ func reconcileGenericWatch[T client.Object, R reconcile.Request](
 func ReconcileWatch[T client.Object, R reconcile.Request](
 	associated types.NamespacedName,
 	associations []commonv1.Association,
-	dynamicRequest *watches.DynamicEnqueueRequest[T, reconcile.Request],
+	dynamicRequest *watches.DynamicEnqueueRequest[T],
 	watchName string,
 	watchedFunc func(association commonv1.Association) types.NamespacedName,
 ) error {
@@ -177,7 +177,7 @@ func ReconcileWatch[T client.Object, R reconcile.Request](
 }
 
 // RemoveWatch removes `watchName` watch from `dynamicRequest`.
-func RemoveWatch[T client.Object, R comparable](dynamicRequest *watches.DynamicEnqueueRequest[T, R], watchName string) {
+func RemoveWatch[T client.Object](dynamicRequest *watches.DynamicEnqueueRequest[T], watchName string) {
 	dynamicRequest.RemoveHandlerForKey(watchName)
 }
 
diff --git a/pkg/controller/common/watches/handler.go b/pkg/controller/common/watches/handler.go
index b039c32a2..f5b653960 100644
--- a/pkg/controller/common/watches/handler.go
+++ b/pkg/controller/common/watches/handler.go
@@ -25,29 +25,29 @@ var (
 
 // HandlerRegistration is the event handler registration that can be added or removed
 // from DynamicEnqueueRequest.
-type HandlerRegistration[T client.Object, R comparable] interface {
+type HandlerRegistration[T client.Object] interface {
 	// Key identifies the transformer
 	Key() string
 	// EventHandler handles CRUD events and turns them into reconcile.Request if relevant.
-	EventHandler() handler.TypedEventHandler[T, R]
+	EventHandler() handler.TypedEventHandler[T, reconcile.Request]
 }
 
 // NewDynamicEnqueueRequest creates a new DynamicEnqueueRequest
-func NewDynamicEnqueueRequest[T client.Object, R reconcile.Request]() *DynamicEnqueueRequest[T, R] {
-	return &DynamicEnqueueRequest[T, R]{
-		registrations: make(map[string]HandlerRegistration[T, R]),
+func NewDynamicEnqueueRequest[T client.Object]() *DynamicEnqueueRequest[T] {
+	return &DynamicEnqueueRequest[T]{
+		registrations: make(map[string]HandlerRegistration[T]),
 	}
 }
 
 // DynamicEnqueueRequest is an EventHandler that allows addition and removal of
 // event handler registrations at runtime allowing dynamic reconciliation based on specific resources.
-type DynamicEnqueueRequest[T client.Object, R comparable] struct {
+type DynamicEnqueueRequest[T client.Object] struct {
 	mutex         sync.RWMutex
-	registrations map[string]HandlerRegistration[T, R]
+	registrations map[string]HandlerRegistration[T]
 }
 
 // AddHandlers adds the new event handlers to this DynamicEnqueueRequest.
-func (d *DynamicEnqueueRequest[T, R]) AddHandlers(handlers ...HandlerRegistration[T, R]) error {
+func (d *DynamicEnqueueRequest[T]) AddHandlers(handlers ...HandlerRegistration[T]) error {
 	for _, h := range handlers {
 		if err := d.AddHandler(h); err != nil {
 			return err
@@ -57,7 +57,7 @@ func (d *DynamicEnqueueRequest[T, R]) AddHandlers(handlers ...HandlerRegistratio
 }
 
 // AddHandler adds a new event handler to this DynamicEnqueueRequest.
-func (d *DynamicEnqueueRequest[T, R]) AddHandler(handler HandlerRegistration[T, R]) error {
+func (d *DynamicEnqueueRequest[T]) AddHandler(handler HandlerRegistration[T]) error {
 	d.mutex.Lock()
 	defer d.mutex.Unlock()
 
@@ -70,19 +70,19 @@ func (d *DynamicEnqueueRequest[T, R]) AddHandler(handler HandlerRegistration[T,
 }
 
 // RemoveHandler removes the handler defined by the transformer.
-func (d *DynamicEnqueueRequest[T, R]) RemoveHandler(handler HandlerRegistration[T, R]) {
+func (d *DynamicEnqueueRequest[T]) RemoveHandler(handler HandlerRegistration[T]) {
 	d.RemoveHandlerForKey(handler.Key())
 }
 
 // RemoveHandlerForKey removes the handler identified by the given key.
-func (d *DynamicEnqueueRequest[T, R]) RemoveHandlerForKey(key string) {
+func (d *DynamicEnqueueRequest[T]) RemoveHandlerForKey(key string) {
 	d.mutex.Lock()
 	defer d.mutex.Unlock()
 	delete(d.registrations, key)
 }
 
 // Registrations returns the list of registered handler names.
-func (d *DynamicEnqueueRequest[T, R]) Registrations() []string {
+func (d *DynamicEnqueueRequest[T]) Registrations() []string {
 	d.mutex.RLock()
 	defer d.mutex.RUnlock()
 	keys := make([]string, 0, len(d.registrations))
@@ -93,10 +93,10 @@ func (d *DynamicEnqueueRequest[T, R]) Registrations() []string {
 }
 
 // DynamicEnqueueRequest implements TypedEventHandler
-var _ handler.TypedEventHandler[client.Object, reconcile.Request] = &DynamicEnqueueRequest[client.Object, reconcile.Request]{}
+var _ handler.TypedEventHandler[client.Object, reconcile.Request] = &DynamicEnqueueRequest[client.Object]{}
 
 // Create is called in response to a create event - e.g. Pod Creation.
-func (d *DynamicEnqueueRequest[T, R]) Create(ctx context.Context, evt event.TypedCreateEvent[T], q workqueue.TypedRateLimitingInterface[R]) {
+func (d *DynamicEnqueueRequest[T]) Create(ctx context.Context, evt event.TypedCreateEvent[T], q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
 	d.mutex.RLock()
 	defer d.mutex.RUnlock()
 	for _, v := range d.registrations {
@@ -105,7 +105,7 @@ func (d *DynamicEnqueueRequest[T, R]) Create(ctx context.Context, evt event.Type
 }
 
 // Update is called in response to an update event -  e.g. Pod Updated.
-func (d *DynamicEnqueueRequest[T, R]) Update(ctx context.Context, evt event.TypedUpdateEvent[T], q workqueue.TypedRateLimitingInterface[R]) {
+func (d *DynamicEnqueueRequest[T]) Update(ctx context.Context, evt event.TypedUpdateEvent[T], q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
 	d.mutex.RLock()
 	defer d.mutex.RUnlock()
 	for _, v := range d.registrations {
@@ -114,7 +114,7 @@ func (d *DynamicEnqueueRequest[T, R]) Update(ctx context.Context, evt event.Type
 }
 
 // Delete is called in response to a delete event - e.g. Pod Deleted.
-func (d *DynamicEnqueueRequest[T, R]) Delete(ctx context.Context, evt event.TypedDeleteEvent[T], q workqueue.TypedRateLimitingInterface[R]) {
+func (d *DynamicEnqueueRequest[T]) Delete(ctx context.Context, evt event.TypedDeleteEvent[T], q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
 	d.mutex.RLock()
 	defer d.mutex.RUnlock()
 	for _, v := range d.registrations {
@@ -124,7 +124,7 @@ func (d *DynamicEnqueueRequest[T, R]) Delete(ctx context.Context, evt event.Type
 
 // Generic is called in response to an event of an unknown type or a synthetic event triggered as a cron or
 // external trigger request - e.g. reconcile Autoscaling, or a Webhook.
-func (d *DynamicEnqueueRequest[T, R]) Generic(ctx context.Context, evt event.TypedGenericEvent[T], q workqueue.TypedRateLimitingInterface[R]) {
+func (d *DynamicEnqueueRequest[T]) Generic(ctx context.Context, evt event.TypedGenericEvent[T], q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
 	d.mutex.RLock()
 	defer d.mutex.RUnlock()
 	for _, v := range d.registrations {
diff --git a/pkg/controller/common/watches/state.go b/pkg/controller/common/watches/state.go
index 805baa39b..d5f8db646 100644
--- a/pkg/controller/common/watches/state.go
+++ b/pkg/controller/common/watches/state.go
@@ -7,26 +7,25 @@ package watches
 import (
 	corev1 "k8s.io/api/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 // NewDynamicWatches creates an initialized DynamicWatches container.
 func NewDynamicWatches() DynamicWatches {
 	return DynamicWatches{
-		ConfigMaps:          NewDynamicEnqueueRequest[*corev1.ConfigMap, reconcile.Request](),
-		Secrets:             NewDynamicEnqueueRequest[*corev1.Secret, reconcile.Request](),
-		Services:            NewDynamicEnqueueRequest[*corev1.Service, reconcile.Request](),
-		Pods:                NewDynamicEnqueueRequest[*corev1.Pod, reconcile.Request](),
-		ReferencedResources: NewDynamicEnqueueRequest[client.Object, reconcile.Request](),
+		ConfigMaps:          NewDynamicEnqueueRequest[*corev1.ConfigMap](),
+		Secrets:             NewDynamicEnqueueRequest[*corev1.Secret](),
+		Services:            NewDynamicEnqueueRequest[*corev1.Service](),
+		Pods:                NewDynamicEnqueueRequest[*corev1.Pod](),
+		ReferencedResources: NewDynamicEnqueueRequest[client.Object](),
 	}
 }
 
 // DynamicWatches contains stateful dynamic watches. Intended as facility to pass around stateful dynamic watches and
 // give each of them an identity.
 type DynamicWatches struct {
-	ConfigMaps          *DynamicEnqueueRequest[*corev1.ConfigMap, reconcile.Request]
-	Secrets             *DynamicEnqueueRequest[*corev1.Secret, reconcile.Request]
-	Services            *DynamicEnqueueRequest[*corev1.Service, reconcile.Request]
-	Pods                *DynamicEnqueueRequest[*corev1.Pod, reconcile.Request]
-	ReferencedResources *DynamicEnqueueRequest[client.Object, reconcile.Request]
+	ConfigMaps          *DynamicEnqueueRequest[*corev1.ConfigMap]
+	Secrets             *DynamicEnqueueRequest[*corev1.Secret]
+	Services            *DynamicEnqueueRequest[*corev1.Service]
+	Pods                *DynamicEnqueueRequest[*corev1.Pod]
+	ReferencedResources *DynamicEnqueueRequest[client.Object]
 }
```

Reply:
@pebrc Ok, this finally sunk into my head this time, and I get this. Sorry for the overdone typing. I can see when these are not needed now. Sorry, generics hadn't 'clicked' yet for me, I guess. I went through and corrected the others that weren't needed as well.